packages/squiggle-lang/src/js/math.ts

```diff
@@ -0,0 +1,91 @@
+type sampleArgs = {
+  probs: number[];
+  size: number;
+};
+
+export function random_sample(dist: number[], args: sampleArgs): number[] {
+  const { probs, size } = args;
+  const sum = probs.reduce((a, b) => a + b);
+
+  let sample: number[] = Array(size);
+  for (let index = 0; index < size; index++) {
+    // Instead of normalizing the whole probability array, we just multiply random by that value.
+    // Might actually be more costly at large sample sizes?
+    let acc = Math.random() * sum;
+    let selection = probs.findIndex((prob) => (acc -= prob) <= 0);
+    // May happen due to fp shenanigans
+    if (selection == -1) selection = size - 1;
+    sample[index] = dist[selection];
+  }
+  return sample;
+}
+
+// Should be replaced by a lookup table if we ever use it too much
+export function factorial(n: number): number {
+  let fct = 1;
+  for (; n > 0; n--) fct *= n;
+  return fct;
+}
+
+type Distribution<P> = {
+  cdf: (args: P, x: number) => number;
+  mean: (args: P) => number;
+  stdev: (args: P) => number;
+  variance: (args: P) => number;
+  quantile: (args: P, p: number) => number;
+};
+
+type ContinuousProbabilityDistribution<P> = Distribution<P> & {
+  pdf: (args: P, x: number) => number;
+};
+
+type DiscreteProbabilityDistribution<P> = Distribution<P> & {
+  pmf: (args: P, x: number) => number;
+};
+
+let bernoulli: DiscreteProbabilityDistribution<number> = {
+  pmf: (p, x) => (x == 0 ? 1 - p : p),
```

Comment:
I prefer to leave it out when I know the types, and hoping that this will be called from the language after checking the arguments, I removed additional type equality check.

I will get it back in place.